mysql_tests/test_declarative.py

```diff
@@ -0,0 +1,305 @@
+    assert hasattr(Model, "col3")
+    assert hasattr(Model.__table__.c, "col1")
+    assert hasattr(Model.__table__.c, "col2")
+    assert hasattr(Model.__table__.c, "col3")
+
+    class Model(db.Model):
+        @db.declared_attr
+        def __tablename__(cls):
+            return "model5"
+
+    assert Model.__table__.name == "model5"
+
+
+async def test_inline_constraints_and_indexes(bind, engine):
+    u = await User.create(nickname="test")
+    us1 = await UserSetting.create(user_id=u.id, setting="skin", value="blue")
+
+    # PrimaryKeyConstraint
+    with pytest.raises(IntegrityError):
+        await UserSetting.create(id=us1.id, user_id=u.id, setting="key1", value="val1")
+
+    # ForeignKeyConstraint
+    with pytest.raises(IntegrityError):
+        await UserSetting.create(user_id=42, setting="key2", value="val2")
+
+    # UniqueConstraint
+    with pytest.raises(IntegrityError):
+        await UserSetting.create(
+            user_id=u.id, setting="skin", value="duplicate-setting"
+        )
+
+    # MySQL doesn't support CheckConstraint
+    # with pytest.raises(CheckViolationError):
+    #     await UserSetting.create(user_id=u.id, setting="key3", value="val3", col1=42)
+
+    # Index
+    status, result = await engine.status(
+        "SHOW INDEXES FROM `gino_user_settings` WHERE Key_name = 'col2_idx'"
+    )
+    assert status == 1
+
+
+async def test_join_t112(engine):
+    class Car(db.Model):
+        __tablename__ = "cars"
+
+        id = db.Column(db.BigInteger(), primary_key=True)
+
+    class Wheel(db.Model):
+        __tablename__ = "wheels"
+
+        id = db.Column(db.BigInteger(), primary_key=True)
+        car_id = db.Column(db.ForeignKey("cars.id"))
+
+    sql = (
+        "SELECT wheels.id, wheels.car_id, cars.id \nFROM wheels "
+        "INNER JOIN cars ON cars.id = wheels.car_id"
+    )
+
+    assert engine.compile(Wheel.join(Car).select())[0] == sql
+
+
+async def test_mixin():
+    class Tracked:
+        created = db.Column(db.DateTime(timezone=True))
+
+        @db.declared_attr
+        def unique_id(cls):
+            return db.Column(db.Integer())
+
+        @db.declared_attr
+        def unique_constraint(cls):
+            return db.UniqueConstraint("unique_id")
+
+        @db.declared_attr
+        def poly(cls):
+            if cls.__name__ == "Thing":
+                return db.Column(db.Unicode())
+
+        @db.declared_attr
+        def __table_args__(cls):
+            if cls.__name__ == "Thing":
+                return (db.UniqueConstraint("poly"),)
+
+    class Audit(Tracked):
+        pass
+
+    class Thing(Audit, db.Model):
+        __tablename__ = "thing"
+
+        id = db.Column(db.Integer, primary_key=True)
+
+    class Another(Audit, db.Model):
+        __tablename__ = "another"
+
+        id = db.Column(db.Integer, primary_key=True)
+
+    assert isinstance(Thing.__table__.c.created, db.Column)
+    assert isinstance(Another.__table__.c.created, db.Column)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/gh/python-gino/gino/pullRequest?prid=5587950)